File: 4_Behavioral_Design_Patterns/4.08_State/Python_Snippet/State.py
from abc import ABC, abstractmethod

class TrafficLightState(ABC):
    @abstractmethod
    def show_light(self):
        pass

    @abstractmethod
    def next_state(self):
        pass

class RedLight(TrafficLightState):
    def show_light(self):
        print("Showing Red Light")

    def next_state(self):
        return YellowLight()

class YellowLight(TrafficLightState):
    def show_light(self):
        print("Showing Yellow Light")

    def next_state(self):
        return GreenLight()

class GreenLight(TrafficLightState):
    def show_light(self):
        print("Showing Green Light")

    def next_state(self):
        return RedLight()

class TrafficLight:
    def __init__(self):
        self.state = RedLight()  # Start with red light

    def change_state(self):
        self.state = self.state.next_state()

    def show_light(self):
        self.state.show_light()

def main():
    light = TrafficLight()

    light.show_light()  # Output: Showing Red Light

    for _ in range(3):
        light.change_state()
        light.show_light()

if __name__ == "__main__":
    main()

File: 4_Behavioral_Design_Patterns/4.08_State/Java_Snippet/State.java
interface TrafficLightState {
    void showLight();

    TrafficLightState nextState();
}

class RedLight implements TrafficLightState {
    @Override
    public void showLight() {
        System.out.println("Showing Red Light");
    }

    @Override
    public TrafficLightState nextState() {
        return new YellowLight();
    }
}

class YellowLight implements TrafficLightState {
    @Override
    public void showLight() {
        System.out.println("Showing Yellow Light");
    }

    @Override
    public TrafficLightState nextState() {
        return new GreenLight();
    }
}

class GreenLight implements TrafficLightState {
    @Override
    public void showLight() {
        System.out.println("Showing Green Light");
    }

    @Override
    public TrafficLightState nextState() {
        return new RedLight();
    }
}

class TrafficLight {
    private TrafficLightState state;

    public TrafficLight() {
        this.state = new RedLight(); // Start with red light
    }

    public void changeState() {
        this.state = state.nextState();
    }

    public void showLight() {
        state.showLight();
    }
}

public class State {
    public static void main(String[] args) {
        TrafficLight light = new TrafficLight();

        light.showLight(); // Output: Showing Red Light

        for (int i = 0; i < 3; i++) {
            light.changeState();
            light.showLight();
        }
    }
}
